File: welcome/pages/layout.go
package pages

import (
	"os"
	"strings"

	"github.com/arthurmelton/KabaOS/welcome/window"
	"github.com/diamondburned/gotk4/pkg/gio/v2"
	"github.com/diamondburned/gotk4/pkg/glib/v2"
	"github.com/diamondburned/gotk4/pkg/gtk/v4"
)

type layout struct {
	code string
	name string
}

type variant struct {
	code   string
	layout string
	name   string
}

var dat, _ = os.ReadFile("/usr/share/X11/xkb/rules/base.lst") // should never fail if we are running on KabaOS
var layouts = getLayouts(string(dat))
var variants = getVariants(string(dat))

var selected = struct {
	a uint
	b uint
}{
	getId("us"),
	0,
}

func Layout(w *window.Welcome, increase bool) {
	screen := gtk.NewBox(gtk.OrientationVertical, 0)

	layoutListSize := uint(0)
	layoutList := gtk.NewStringList([]string{})

	layout := gtk.NewDropDown(&layoutList.ListModel, nil)

	language := gtk.NewDropDownFromStrings(getNames())
	language.Connect("notify::selected", func() {
		variantNames := getVariantNames(layouts[language.Selected()].code)
		layoutList.Splice(0, layoutListSize, variantNames)
		layoutListSize = uint(len(variantNames))
		selected.a = language.Selected()
		layout.SetSelected(0)
	})

	layout.Connect("notify::selected", func() {
		updateLayout(getLayoutString(layouts[language.Selected()].code, getVariantNames(layouts[language.Selected()].code)[layout.Selected()]))
		selected.b = layout.Selected()
	})

	tmpb := selected.b
	language.SetSelected(selected.a)
	layout.SetSelected(tmpb)

	screen.Append(language)
	screen.Append(layout)

	w.Window.SetTitle("Select a keyboard layout")
	w.Window.SetChild(screen)
}

func getLayouts(base string) []layout {
	inside, returns := false, []layout{}
	for _, line := range strings.Split(base, "\n") {
		if line == "" {
			inside = false
		}

		if inside {
			feilds := strings.Fields(line)

			returns = append(returns, layout{
				code: feilds[0],
				name: strings.Join(feilds[1:], " "),
			})
		}

		if line == "! layout" {
			inside = true
		}
	}
	return returns
}

func getVariants(base string) []variant {
	inside, returns := false, []variant{}
	for _, line := range strings.Split(base, "\n") {
		if line == "" {
			inside = false
		}

		if inside {
			feilds := strings.Fields(line)

			returns = append(returns, variant{
				code:   feilds[0],
				layout: feilds[1][:len(feilds[1])-1],
				name:   strings.Join(feilds[2:], " "),
			})
		}

		if line == "! variant" {
			inside = true
		}
	}
	return returns
}

func getId(code string) uint {
	for i := range layouts {
		if layouts[i].code == code {
			return uint(i)
		}
	}
	return 0
}

func getNames() []string {
	result := []string{}
	for i := range layouts {
		result = append(result, layouts[i].name)
	}
	return result
}

func getLayoutCode(search string) string {
	for i := range layouts {
		if layouts[i].name == search {
			return layouts[i].code
		}
	}
	return "" // should never happen
}

func getVariantNames(code string) []string {
	result := []string{"Default"}
	for i := range variants {
		if variants[i].layout == code {
			result = append(result, variants[i].name)
		}
	}
	return result
}

func getLayoutString(layout string, variant string) string {
	for i := range variants {
		if variants[i].name == variant {
			return layout + "+" + variants[i].code
		}
	}
	return layout
}

func updateLayout(name string) {
	g := gio.NewSettings("org.gnome.desktop.input-sources")
	g.SetValue("sources", glib.NewVariantArray(glib.NewVariantType("(ss)"), []*glib.Variant{glib.NewVariantTuple([]*glib.Variant{glib.NewVariantString("xkb"), glib.NewVariantString(name)})}))
}

File: welcome/src/pages/layout.zig
const std = @import("std");
const allocator = @import("../main.zig").allocator;
const window = @import("../main.zig").global_window;
const c = @cImport({
    @cInclude("gtk/gtk.h");
    @cInclude("gio/gio.h");
    @cInclude("glib.h");
    @cInclude("adwaita.h");
});

var first = true;

const layouts = struct {
    var code = std.ArrayList([]u8).init(allocator);
    var name = std.ArrayList([]u8).init(allocator);
};
const variants = struct {
    var code = std.ArrayList([]u8).init(allocator);
    var v_code = std.ArrayList([]u8).init(allocator);
    var name = std.ArrayList([]u8).init(allocator);
};

var layoutList: ?*c.GtkStringList = null;
var layoutLength: u32 = 0;
var layoutDD: ?*c.GtkWidget = null;
var languageDD: ?*c.GtkWidget = null;
var layoutHolder: ?*c.GtkWidget = null;
var screenLayout: ?*c.GtkWidget = null;

pub fn page(_: bool) void {
    if (first) {
        first = false;
        layouts_set() catch @panic("Something is wrong with your xkb base.lsl");
    }

    const languageList = c.gtk_string_list_new(null);
    layoutList = c.gtk_string_list_new(null);
    c.gtk_string_list_append(layoutList, "Default");

    // I am too sleepy to get actually get the arraylist strings to convert nicely
    for (layouts.name.items) |x| {
        c.gtk_string_list_append(languageList, x.ptr);
    }
    languageDD = c.adw_combo_row_new();
    c.adw_combo_row_set_model(@ptrCast(languageDD), @ptrCast(languageList));
    c.g_object_set(languageDD, "title", "Language", c.NULL);

    _ = c.g_signal_connect_data(languageDD, "notify::selected", @ptrCast(&struct {
        fn f(_: *c.GtkApplication, _: c.gpointer) callconv(.C) void {
            c.gtk_string_list_splice(layoutList, 1, layoutLength, null);
            layoutLength = 0;
            const selected = layouts.code.items[c.adw_combo_row_get_selected(@ptrCast(languageDD))];
            for (variants.name.items, variants.code.items) |x, i| {
                if (std.mem.eql(u8, i, selected)) {
                    c.gtk_string_list_append(layoutList, x.ptr);
                    layoutLength += 1;
                }
            }
            window.pages.layout.languageSelected = @intCast(c.adw_combo_row_get_selected(@ptrCast(languageDD)));
            c.adw_combo_row_set_selected(@ptrCast(layoutDD), 0);
        }
    }.f), null, null, 0);

    layoutDD = c.adw_combo_row_new();
    c.adw_combo_row_set_model(@ptrCast(layoutDD), @ptrCast(layoutList));
    c.g_object_set(layoutDD, "title", "Layout", c.NULL);

    _ = c.g_signal_connect_data(layoutDD, "notify::selected", @ptrCast(&struct {
        fn f(_: *c.GtkApplication, _: c.gpointer) callconv(.C) void {
            const selectedLanguage = c.adw_combo_row_get_selected(@ptrCast(languageDD));
            const selectedVariant = c.adw_combo_row_get_selected(@ptrCast(layoutDD));
            if (selectedVariant == 0) {
                layout_update(layouts.code.items[selectedLanguage][0..2]);
            } else {
                const currentLanguage = layouts.code.items[selectedLanguage];
                var currentVariant: ?[]u8 = null;
                var index: usize = 0;
                for (variants.code.items, variants.v_code.items) |k, v| {
                    if (std.mem.eql(u8, k, currentLanguage)) {
                        if (index == selectedVariant - 1) {
                            currentVariant = v;
                            break;
                        }
                        index += 1;
                    }
                }
                var result = allocator.alloc(u8, currentLanguage.len + currentVariant.?.len) catch @panic("Failed to initilize string");
                defer allocator.free(result);

                std.mem.copyForwards(u8, result[0..], currentLanguage);
                result[currentLanguage.len - 1] = '+';
                std.mem.copyForwards(u8, result[currentLanguage.len..], currentVariant.?);
                layout_update(result);
            }
            window.pages.layout.variantSelected = @intCast(selectedVariant);
        }
    }.f), null, null, 0);

    layoutHolder = c.gtk_list_box_new();

    c.gtk_list_box_set_selection_mode(@ptrCast(layoutHolder), c.GTK_SELECTION_NONE);
    c.gtk_widget_add_css_class(@ptrCast(layoutHolder), "boxed-list");

    c.gtk_list_box_append(@ptrCast(layoutHolder), languageDD);
    c.gtk_list_box_append(@ptrCast(layoutHolder), layoutDD);

    screenLayout = c.gtk_box_new(c.GTK_ORIENTATION_VERTICAL, 0);
    c.gtk_box_set_spacing(@ptrCast(screenLayout), 18);
    c.gtk_widget_set_margin_top(screenLayout, 24);
    c.gtk_widget_set_margin_bottom(screenLayout, 24);
    c.gtk_widget_set_margin_start(screenLayout, 12);
    c.gtk_widget_set_margin_end(screenLayout, 12);
    c.gtk_box_append(@ptrCast(screenLayout.?), layoutHolder);

    const tmpSelected = window.pages.layout.variantSelected;

    if (window.pages.layout.languageSelected == -1) {
        for (layouts.code.items, 0..) |code, i| {
            if (std.mem.eql(u8, code[0..2], "us")) {
                c.adw_combo_row_set_selected(@ptrCast(languageDD), @intCast(i));
                break;
            }
        }
    } else {
        c.adw_combo_row_set_selected(@ptrCast(languageDD), @intCast(window.pages.layout.languageSelected));
    }

    c.adw_combo_row_set_selected(@ptrCast(layoutDD), @intCast(tmpSelected));

    c.gtk_window_set_child(@ptrCast(window.window.?), screenLayout);
    c.gtk_window_set_title(@ptrCast(window.window.?), "Select a keyboard layout");
}

fn layouts_set() !void {
    const file = try std.fs.openFileAbsolute("/usr/share/X11/xkb/rules/base.lst", .{});
    defer file.close();

    var buf_reader = std.io.bufferedReader(file.reader());
    const reader = buf_reader.reader();

    var line = std.ArrayList(u8).init(allocator);
    defer line.deinit();

    const writer = line.writer();

    const states = enum { BeforeLayout, Layout, AfterLayout, Variant, AfterVariant };
    var state: states = states.BeforeLayout;

    while (reader.streamUntilDelimiter(writer, '\n', null)) {
        defer line.clearRetainingCapacity();

        if (std.mem.eql(u8, line.items, "")) {
            if (state == states.Layout) {
                state = states.AfterLayout;
            } else if (state == states.Variant) {
                state = states.AfterVariant;
            }
        }

        var index: usize = 0;

        var current = std.ArrayList(u8).init(allocator);
        defer current.deinit();

        if (state == states.Layout) {
            while (line.items[index] == ' ') {
                index += 1;
            }
            while (line.items[index] != ' ') {
                try current.append(line.items[index]);
                index += 1;
            }

            try current.append(0);
            try layouts.code.append((try current.clone()).items);
            current.shrinkAndFree(0);

            while (line.items[index] == ' ') {
                index += 1;
            }
            while (line.items.len > index) {
                try current.append(line.items[index]);
                index += 1;
            }

            try current.append(0);
            try layouts.name.append((try current.clone()).items);
            current.shrinkAndFree(0);
        } else if (state == states.Variant) {
            while (line.items[index] == ' ') {
                index += 1;
            }
            while (line.items[index] != ' ') {
                try current.append(line.items[index]);
                index += 1;
            }

            try current.append(0);
            try variants.v_code.append((try current.clone()).items);
            current.shrinkAndFree(0);

            while (line.items[index] == ' ') {
                index += 1;
            }
            while (line.items[index] != ':') {
                try current.append(line.items[index]);
                index += 1;
            }
            index += 1;

            try current.append(0);
            try variants.code.append((try current.clone()).items);
            current.shrinkAndFree(0);

            while (line.items[index] == ' ') {
                index += 1;
            }
            while (line.items.len > index) {
                try current.append(line.items[index]);
                index += 1;
            }

            try current.append(0);
            try variants.name.append((try current.clone()).items);
            current.shrinkAndFree(0);
        }

        if (state == states.BeforeLayout and std.mem.eql(u8, line.items, "! layout")) {
            state = states.Layout;
        } else if (state == states.AfterLayout and std.mem.eql(u8, line.items, "! variant")) {
            state = states.Variant;
        }
    } else |err| switch (err) {
        error.EndOfStream => {},
        else => return err,
    }

    // Remove "custom"
    _ = layouts.name.pop();
    _ = layouts.code.pop();
}

fn layout_update(name: []const u8) void {
    const settings = c.g_settings_new("org.gnome.desktop.input-sources");
    _ = c.g_settings_set_value(settings, "sources", c.g_variant_new_array(c.g_variant_type_new("(ss)"), &[_]?*c.GVariant{c.g_variant_new_tuple(&[_]?*c.GVariant{ c.g_variant_new_string("xkb"), c.g_variant_new_string(@ptrCast(name)) }, 2)}, 1));
}
